### PR TITLE
Implement OAuth

### PR DESCRIPTION
**Pull request by kashizui**
Incomplete pull request. Plan is to:
1. ~~Flesh out schemas for users and oauth tables~~
2. ~~Merge with @klopyrev's bottle patch~~
3. ~~**Port Flask-OAuthlib to bottle**~~
4. ~~Implement ORM objects for the User and OAuth models~~
5. ~~Implement the corresponding BundleModel methods for fetching and saving these objects~~
6. ~~Scaffold out OAuth2 provider app in Bottle~~
7. ~~Implement Resource Owner Credential Flow~~
8. ~~Test token retrieval~~
9. ~~Implement remaining authorization flows~~
10. ~~Implement proper password hashing~~
11. ~~Fix client_id~~
12. ~~Build authorize and login views~~
13. ~~Do full end-to-end tests of the authorization flows (√authcode, √password, √refreshtoken, √implicit)~~
14. ~~make cookies secure~~
15. ~~Set up and test authorization (bearer token) check~~

### CODE REVIEW

**Comment by kashizui** [comment]
Successful test of the Resource Owner Password Credentials Grant flow!

screen shot 2016-02-15 at 5 37 27 pm

**Comment by percyliang** [comment]
Cool!

**Comment by kashizui** [comment]
screen shot 2016-02-18 at 11 06 53 pm
screen shot 2016-02-18 at 11 07 14 pm

**Comment by kashizui** [comment]
This is nearly ready to go, just needs some cleanup and the user signup and email confirmation stuff still needs to be implemented (perhaps can be saved for another pull request, if this one can be merged soon enough).

**Comment by kashizui** [comment]
@percyliang ready for review, or can just be merged now since it's independent of the existing server/CLI code.

File: codalab/model/tables.py [hunk 1]
```diff
@@ -216,7 +217,16 @@
 
   # Basic information
   Column('user_id', String(63), nullable=False),
-  Column('user_name', String(63), nullable=False),  # Mirrors the OAuth server (eventually move it here)
+  Column('user_name', String(63), nullable=False),
+  Column('email', String(254), nullable=False),  # Length of 254 to be compliant with RFC3696/5321
+  Column('last_login', DateTime),  # Null if user has never logged in
```
Comment (percyliang) [inline codalab/model/tables.py:222]: Might as well add the fields affiliation and url

File: codalab/model/tables.py [hunk 2]
```diff
@@ -226,5 +236,62 @@
 
   Index('user_user_id_index', 'user_id'),
   Index('user_user_name_index', 'user_name'),
+  UniqueConstraint('user_id', name='uix_1'),
+  sqlite_autoincrement=True,
+)
+
+# Stores verification keys
```
Comment (percyliang) [inline codalab/model/tables.py:247]: What is this used for?

File: codalab/server/oauth2_provider.py [hunk 3]
```diff
@@ -0,0 +1,870 @@
+# coding: utf-8
+"""
```
Comment (percyliang) [inline codalab/server/oauth2_provider.py:2]: Comment on what was changed from the original file.

**Comment by percyliang** [comment]
LGTM!

I think having some high-level description of what the different components and how they fit together are would be helpful.  Should we put this in one of the py files?

File: codalab/server/rest_server.py [hunk 4]
```diff
@@ -73,10 +82,20 @@ def run_rest_server(manager, debug, num_workers):
     host = manager.config['server']['rest_host']
     port = manager.config['server']['rest_port']
 
-    install(SaveEnvironmentPlugin(manager.model(), manager.bundle_store()))
+    install(SaveEnvironmentPlugin(manager.model(), manager.bundle_store(), manager.config))
     install(CheckJsonPlugin())
     install(LoggingPlugin())
 
+    # Import views
+    import codalab.rest.login
```
Comment (klopyrev) [inline codalab/server/rest_server.py]: Why are you importing here?

File: codalab/rest/login.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        if redirect_uri:
+            return redirect(redirect_uri)
+        else:
+            return "<p>Successfully signed into CodaLab.</p>"
+    else:
+        return template("login", redirect_uri=redirect_uri, error="Login/password did not match.")
+
+
+# The other way to do this is to write a Plugin and add it to the "apply" param to the authorize view function
+def require_login(callback):
```
Comment (klopyrev) [inline codalab/rest/login.py]: I already have the authorization decorator... That's supposed to do what this does.

File: codalab/server/rest_server.py [hunk 6]
```diff
@@ -73,10 +82,20 @@ def run_rest_server(manager, debug, num_workers):
...
+    import codalab.rest.users
+    from codalab.rest.oauth2 import oauth2_app
+    import codalab.rest.example
+
+    # Mount OAuth2 endpoints
+    oauth2_app.install(SaveEnvironmentPlugin(manager.model(), manager.bundle_store(), manager.config))
```
Comment (klopyrev) [inline codalab/server/rest_server.py]: Why does this need to be an app? I have an issue with the fact that you're installing this plugin again.

File: codalab/lib/codalab_manager.py [hunk 7]
```diff
@@ -232,6 +241,34 @@ def init_config(self, dry_run=False):
 
         return config
 
+    def init_logger(self):
```
Comment (klopyrev) [inline codalab/lib/codalab_manager.py]: This change should go into a separate pull request, since it's unrelated to the OAuth2 change.

File: codalab/lib/jsonapi.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+"""
```
Comment (klopyrev) [inline codalab/lib/jsonapi.py]: This file is empty.

File: codalab/lib/server_util.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ Status API Training Shop Blog About Pricing
+"""
+
+# BEGIN ADAPTED FROM werkzeug.utils #
+
+
+class cached_property(property):
```
Comment (klopyrev) [inline codalab/lib/server_util.py]: See cached in codalab_manager.py

File: codalab/model/tables.py (hunk 2 above)
Comment (kashizui) [inline codalab/model/tables.py:247]: This is for the email verification process! Adapted (a simplified version of) this schema from the original Django tables.
